pyout/common.py

```diff
@@ -0,0 +1,395 @@
+                callables.append((columns, fn))
+
+        for column, value in to_add:
+            row[column] = value
+        for multi_columns in to_delete:
+            del row[multi_columns]
+
+        return callables
+
+    # Input-specific getters.  These exist as their own methods so that they
+    # can be wrapped in a callable and delayed.
+
+    def getter_dict(self, row, column):
+        return row.get(column, self.nothings[column])
+
+    def getter_seq(self, row, column):
+        col_to_idx = {c: idx for idx, c in enumerate(self._columns)}
+        return row[col_to_idx[column]]
+
+    def getter_attrs(self, row, column):
+        return getattr(row, column, self.nothings[column])
+
+
+def _safe_get(mapping, key, default=None):
+    try:
+        return mapping.get(key, default)
+    except AttributeError:
+        return default
+
+
+class StyleFields(object):
+    """Generate Fields based on the specified style and processors.
+
+    Parameters
+    ----------
+    style : dict
+        A style that follows the schema defined in pyout.elements.
+    procgen : StyleProcessors instance
+        This instance is used to generate the fields from `style`.
+    """
+
+    _header_attributes = {"align", "width"}
+
+    def __init__(self, style, procgen):
+        self.init_style = style
+        self.procgen = procgen
+
+        self.style = None
+        self.columns = None
+        self.autowidth_columns = {}
+
+        self.fields = None
+
+    def build(self, columns):
+        """Build the style and fields.
+
+        Parameters
+        ----------
+        columns : list of str
+            Column names.
+        """
+        self.columns = columns
+        default = dict(elements.default("default_"),
+                       **_safe_get(self.init_style, "default_", {}))
+        self.style = elements.adopt({c: default for c in columns},
+                                    self.init_style)
+
+        hstyle = None
+        if self.init_style is not None and "header_" in self.init_style:
+            hstyle = {}
+            for col in columns:
+                cstyle = {k: v for k, v in self.style[col].items()
+                          if k in self._header_attributes}
+                hstyle[col] = dict(cstyle, **self.init_style["header_"])
+
+        # Store special keys in _style so that they can be validated.
+        self.style["default_"] = default
+        self.style["header_"] = hstyle
+        self.style["separator_"] = _safe_get(self.init_style, "separator_",
+                                             elements.default("separator_"))
+        elements.validate(self.style)
+        self._setup_fields()
+
+    def _setup_fields(self):
+        self.fields = {}
+        for column in self.columns:
+            cstyle = self.style[column]
+
+            core_procs = []
+            style_width = cstyle["width"]
+            is_auto = style_width == "auto" or _safe_get(style_width, "auto")
```

Comment:
confused... if I specify `width=5`, above would make is_auto True?  so, it could be False only if I set it to False? (0 would make it still non-False)

Reply:
The default is None here, so is_auto will be False for style_width=5.